job-runner/chain/traverser.go

```diff
@@ -377,3 +395,85 @@ func (t *traverser) sendJL(pJob proto.Job, err error) {
 		jLogger.Errorf("problem sending job log (%#v) to the RM: %s", jl, err)
 	}
 }
+
+// retrySequence retries a sequence by doing the following
+//   1. Mark the failed job and all jobs before it in the sequence to PENDING
+//   state, so that they are ran again on sequence retry. This is OK to do
+//   since all jobs are idempotent.
+//   2. Decrement the number of jobs that were previously ran in the sequence
+//   from the chain job count.
+//   3. Increment retry count for the sequence
+//   4. Enqueue the first job in the sequence to be ran again. It'll
+//   subsequently enqueue the remaining jobs in the sequence.
+func (t *traverser) retrySequence(failedJob proto.Job) error {
+	sequenceStartJob := t.chain.SequenceStartJob(failedJob)
+
+	jobIdsToRetry := t.jobIdsToRetry(sequenceStartJob.Id)
+	// jobIdsToRetry is now a list of jobs we have to undo
+	for _, jobId := range jobIdsToRetry {
+		t.chain.SetJobState(jobId, proto.STATE_PENDING)
+	}
+
+	// Decrement number of jobs we will attempt to retry
+	t.chain.n -= uint(len(jobIdsToRetry))
+
+	// Increment retry count for this sequence
+	t.chain.IncrementSequenceRetryCount(failedJob)
+
+	// Enqueue first job in sequence
+	t.runJobChan <- sequenceStartJob
+
+	return nil
+}
+
+// jobIdsToRetry does a breadth first search to find all jobs whose state is not PENDING, which means the state is either COMPLETE or FAILED. This is a list of jobs that need to be retried.
+func (t *traverser) jobIdsToRetry(sequenceStartJobId string) []string {
+	// visited is a list of job ids visited
+	var visited []string
+	// frontier is a list of job ids to visit
+	var frontier []string
```

Comment:
s/frontier/toVisit/  or /pending/? Unless "frontier" is normal terminology for this type of code.

Reply:
frontier is normal terminology for breadth first search and the same verbiage used in a different part in this repo. 

https://github.com/square/spincycle/blob/1e3a3845f4c6fdc0ca9974fac2aad276673f32d8/request-manager/grapher/graph.go#L241

Reply:
I've renamed it to `toVisit`. although frontier is normal terminology it did confuse you and it confused me at first until i googled it. so we might as well go for the more intuitive variable name since it has the same implication.